packages/foundry/contracts/hooks/SwapDiscountHook/DiscountCampaign.sol

```diff
@@ -0,0 +1,212 @@
+     * @param user The address of the user receiving the discount.
+     * @param swappedAmount The amount that was swapped.
+     * @param timeOfSwap The timestamp of when the swap occurred.
+     */
+    function updateUserDiscountMapping(
+        bytes32 campaignID,
+        uint256 tokenId,
+        address user,
+        uint256 swappedAmount,
+        uint256 timeOfSwap
+    ) external override {
+        require(msg.sender == address(_swapHook), "Unauthorized");
+
+        userDiscountMapping[tokenId] = UserSwapData({
+            campaignID: campaignID,
+            userAddress: user,
+            campaignAddress: address(this),
+            swappedAmount: swappedAmount,
+            timeOfSwap: timeOfSwap,
+            hasClaimed: false
+        });
+    }
+
+    /**
+     * @notice Claims rewards for a specific token ID.
+     * @dev Transfers the reward to the user associated with the token and marks the token as claimed.
+     *      Reverts if the reward amount is zero or if the total rewards have been distributed.
+     * @param tokenID The ID of the token for which the claim is made.
+     */
+    function claim(uint256 tokenID) public checkAndAuthorizeTokenId(tokenID) nonReentrant {
+        UserSwapData memory userSwapData = userDiscountMapping[tokenID];
+        uint256 reward = _getClaimableRewards(userSwapData);
+
+        if (reward == 0) {
+            revert RewardAmountExpired();
+        }
+
+        tokenRewardDistributed += reward;
+        _maxBuy -= reward;
+        userDiscountMapping[tokenID].hasClaimed = true;
+        TransferHelper.safeTransfer(campaignDetails.rewardToken, userSwapData.userAddress, reward);
+        _updateDiscount();
+    }
+
+    /**
+     * @notice Internal function to calculate the claimable reward for a given token ID.
+     * @dev The reward is calculated based on the swapped amount and discount rate.
+     * @param tokenID The ID of the token for which to calculate the reward.
+     * @return claimableReward The amount of reward that can be claimed.
+     */
+    function getClaimableReward(uint256 tokenID) external view returns (uint256 claimableReward) {
+        UserSwapData memory userSwapData = userDiscountMapping[tokenID];
+        return _getClaimableRewards(userSwapData);
+    }
+
+    /**
+     * @notice Overloaded version of _getClaimableRewards that takes UserSwapData as input.
+     * @param userSwapData The UserSwapData struct containing the necessary information for calculating rewards.
+     * @return claimableReward The amount of reward that can be claimed.
+     */
+    function _getClaimableRewards(UserSwapData memory userSwapData) private view returns (uint256 claimableReward) {
+        uint256 swappedAmount = userSwapData.swappedAmount;
+
+        // Calculate claimable reward based on the swapped amount and discount rate
+        if (swappedAmount <= _maxBuy) {
+            claimableReward = (swappedAmount * campaignDetails.discountRate) / 100e18;
```

Comment:
The rest of the system uses FixedPoint math (i.e., 1 = 1e18). For instance, a 20% rate would be 20e16. So `claimableReward = swappedAmount.mulDown(campaignDetails.discountRate)`.